### PR TITLE
Thresholded metrics

### PR DESCRIPTION
**Pull request by diegokolena**
Associated SDK Server Modifications Complementing These Client Updates:
- https://github.com/kolenaIO/shipyard/pull/1710

### What change does this PR introduce and why?
Update way we sent thresholded metrics to the Kolena Backend. 

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

File: kolena/workflow/test_run.py [hunk 1]
```diff
@@ -428,6 +457,28 @@ def _upload_test_sample_metrics(
         )
         krequests.raise_for_status(res)
 
+        if len(thresholded_metrics) > 0:
+            log.info("Uploading thresholded metrics")
+            df = pd.DataFrame(thresholded_metrics, columns=["test_sample", "metrics"])
+            df_validated = MetricsDataFrame(validate_df_schema(df, MetricsDataFrameSchema, trusted=True))
+            df_serializable = df_validated.as_serializable()
+            log.info(df_serializable)
```
Comment (munkyshi) [inline kolena/workflow/test_run.py]: I assume this should be removed? We wouldn't want to log an entire data frame.

File: kolena/_api/v1/generic.py [hunk 2]
```diff
@@ -68,6 +68,7 @@ class Path(str, Enum):
         LOAD_INFERENCES = "/generic/test-run/load-test-sample-inferences"
         UPLOAD_INFERENCES = "/generic/test-run/upload-inferences"
         UPLOAD_TEST_SAMPLE_METRICS = "/generic/test-run/upload-test-sample-metrics"
+        UPLOAD_TEST_SAMPLE_METRICS_THRESHOLDED = "/generic/test-run/upload-test-sample-metrics-thresholded"
```
Comment (munkyshi) [inline kolena/_api/v1/generic.py]: nit: the type is `ThresholdedMetrics`, correct? We should be consistent with how the term is used.
```suggestion
        UPLOAD_TEST_SAMPLE_THRESHOLDED_METRICS = "/generic/test-run/upload-test-sample-thresholded-metrics"
```

File: kolena/_api/v1/generic.py [hunk 3]
```diff
@@ -115,6 +116,12 @@ class UploadTestSampleMetricsRequest(BatchedLoad.WithLoadUUID):
         test_case_id: Optional[int]
         configuration: Optional["TestRun.EvaluatorConfiguration"]
 
+    @dataclass(frozen=True)
+    class UploadTestSampleMetricsThresholdedRequest(BatchedLoad.WithLoadUUID):
```
Comment (munkyshi) [inline kolena/_api/v1/generic.py]:
```suggestion
    class UploadTestSampleThresholdedMetricsRequest(BatchedLoad.WithLoadUUID):
```

**Comment by munkyshi** [comment]
Is there a corresponding sdk-server PR to review alongside this?

File: kolena/workflow/test_run.py [hunk 4]
```diff
@@ -401,6 +401,34 @@ def _iter_test_samples_batch(
             df_class=TestSampleDataFrame,
         )
 
+    def _extract_thresholded_metrics(
```
Comment (munkyshi) [inline kolena/workflow/test_run.py:410]: Any chance this can be unit tested?

File: kolena/_api/v1/generic.py (hunk 2 above)
Comment (diegokolena) [reply kolena/_api/v1/generic.py]: updated

File: kolena/workflow/test_run.py (hunk 1 above)
Comment (diegokolena) [reply kolena/workflow/test_run.py]: removed

File: kolena/workflow/test_run.py [hunk 5]
```diff
@@ -401,6 +401,34 @@ def _iter_test_samples_batch(
...
+        removed_items = []
+
+        for first_record, second_record in records:
+            # Track keys to remove from second_record after iteration
+            keys_to_remove = []
+
+            for key, value in second_record.items():
+                if isinstance(value, list):
+                    for item in value:
+                        if item.get("data_type") == "METRICS/THRESHOLDED":
```
Comment (ly-kolena) [inline kolena/workflow/test_run.py]: Is `item` guaranteed to be dictionary here?

Comment (diegokolena) [reply kolena/workflow/test_run.py]: Yes, item will be the fields define on the class.

File: kolena/workflow/test_run.py (hunk 4 above)
Comment (diegokolena) [reply kolena/workflow/test_run.py:410]: unit test was added

Comment (ly-kolena) [reply kolena/workflow/test_run.py:410]: Is the plan to add integration test after backend PR?

File: kolena/workflow/test_run.py (hunk 5 above)
Comment (ly-kolena) [reply kolena/workflow/test_run.py]: `second_record` is a general test sample metrics object/dictionary, so value could be a list of str, float, etc., right?

Comment (diegokolena) [reply kolena/workflow/test_run.py]: yes you are right, I was just thinking on the cases of thresholded metrics. I updated the code and add new scenario on unit test to cover that case.

File: tests/unit/workflow/test_test_run.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+def test__when_no_threshold_metric_classes_non_dict_item(mock_test_run_init, mock_krequests_put) -> None:
+    # Create a TestRun instance
+    test_run = TestRun()
+
+    # Define the input records with a non-dict item in the list
+    records = [
+        (
+            {"locator": "s3://bucket/image1.jpg", "data_type": "TEST_SAMPLE/IMAGE"},
+            {
+                "metric_1": ["non-dict-item", 1, 12.0, True, {"data_type": "METRICS/REGULAR"}],
```
Comment (munkyshi) [inline tests/unit/workflow/test_test_run.py:78]: I'm confused, is `METRICS/REGULAR` a thing now?

File: kolena/workflow/test_run.py [hunk 7]
```diff
@@ -401,14 +407,44 @@ def _iter_test_samples_batch(
             df_class=TestSampleDataFrame,
         )
 
+    def _extract_thresholded_metrics(
+        self,
+        records: List[Tuple[Dict[str, Any], Dict[str, Any]]],
+    ) -> Tuple[List[Tuple[Dict[str, Any], Dict[str, Any]]], List[Tuple[Dict[str, Any], List[Dict[str, Any]]]]]:
+        updated_records = []
+        removed_items = []
```
Comment (munkyshi) [inline kolena/workflow/test_run.py]:
```suggestion
        standard_metrics = []
        thresholded_metrics = []
```

nit: It's really confusing to read this and wonder what `removed_items` is, especially given that the method is called `_extract_thresholded_metrics` -- suggest updating in this method + in the unit tests

**Review by munkyshi — APPROVED** [review]
Added one naming nit, but I don't want to be the reason this PR is blocked anymore. Sorry it took so long to get to this. I'll take a look at https://github.com/kolenaIO/kolena/pull/375 next

File: tests/unit/workflow/test_test_run.py (hunk 6 above)
Comment (diegokolena) [reply tests/unit/workflow/test_test_run.py:78]: No, I was just using something different than METRICS/THRESHOLDED

File: kolena/workflow/test_run.py (hunk 7 above)
Comment (diegokolena) [reply kolena/workflow/test_run.py]: updated with this suggestions

File: tests/unit/workflow/test_test_run.py (hunk 6 above)
Comment (munkyshi) [reply tests/unit/workflow/test_test_run.py:78]: Does `METRICS/REGULAR` make a nested data object valid now?

Comment (diegokolena) [reply tests/unit/workflow/test_test_run.py:78]: There isn't strict validation for that field; it only processes when the data_type is METRICS/THRESHOLDED.

**Pull request merged**